README.md
# ACiD

Artificial Circuit Designer, a 
[circus](https://github.com/augustunderground/circus) clown.

## Dependencies

- [HaskTorch](https://github.com/hasktorch/hasktorch)
- [MLFlow](https://github.com/AugustUnderground/mlflow-hs)
- [Circus](https://github.com/AugustUnderground/circus)

## Command Line Interface

Run a training on `op2` in `xh035` with a `TD3` agent and `HER` buffer in the
`Electric` design space:

```bash
$ stack exec -- acid-exe -l TD3 -b HER -i op2 -p xh035 -v 0 -s elec
```

Here is the short help for the CLI and available options:

```bash
ACiD

Usage: acid-exe [-l|--algorithm ALGORITHM] [-b|--buffer BUFFER]
                [-H|--circus-host HOST] [-P|--circus-port PORT] [-i|--ace ID]
                [-p|--pdk PDK] [-s|--space SPACE] [-v|--var VARIANT]
                [-f|--path FILE] [-T|--tracking-host HOST]
                [-R|--tracking-port PORT] [-m|--mode MODE]
  Artificial Circuit Designer / Circus Clown

Available options:
  -l,--algorithm ALGORITHM DRL Algorithm. One of td3 (default), sac, ppo
                           (default: "td3")
  -b,--buffer BUFFER       Replay Buffer Type. One of her (default: "HER")
  -H,--circus-host HOST    Circus server host address (default: "localhost")
  -P,--circus-port PORT    Circus server port (default: "6007")
  -i,--ace ID              ACE OP ID. One of op1, op2 (default), op8
                           (default: "op2")
  -p,--pdk PDK             ACE Backend. One of xh018, xh035 (default)
                           (default: "xh035")
  -s,--space SPACE         Design / Action space. elec (default) or geom
                           (default: "elec")
  -v,--var VARIANT         Circus Environment Variant. 0 = goal (default), 1 =
                           non-goal (default: "0")
  -f,--path FILE           Base Path for Model Checkpoint. Default is ./models
                           (default: "./models")
  -T,--tracking-host HOST  MLFlow tracking server host address
                           (default: "localhost")
  -R,--tracking-port PORT  MLFlow tracking server port (default: "6008")
  -m,--mode MODE           Run Mode. One of Train (default), Continue, Evaluate
                           (default: "Train")
  -h,--help                Show this help text
```

## TODO

- [X] TD3
- [ ] SAC
- [ ] PPO
- [X] HER
- [ ] PER
- [ ] ERE
- [ ] Hyper Parameters as YAML


ChangeLog.md
# Changelog for acid

## Unreleased changes


app/Main.hs
{-# OPTIONS_GHC -Wall #-}

module Main where

import Options.Applicative

import Lib
import ACiD

main :: IO ()
main =  execParser opts >>= run
  where
    desc = "Artificial Circuit Designer / Circus Clown"
    opts = info (args <**> helper) 
                (fullDesc <> progDesc desc <> header "ACiD")

-- | Command Line Argument Parser
args :: Parser Args
args = Args <$> strOption ( long "algorithm" 
                         <> short 'l'
                         <> metavar "ALGORITHM" 
                         <> showDefault 
                         <> value "td3"
                         <> help "DRL Algorithm. One of td3 (default), sac, ppo" )
            <*> strOption ( long "buffer" 
                         <> short 'b'
                         <> metavar "BUFFER" 
                         <> showDefault 
                         <> value "HER"
                         <> help "Replay Buffer Type. One of her" )
            <*> strOption ( long "circus-host" 
                         <> short 'H'
                         <> metavar "HOST" 
                         <> showDefault 
                         <> value "localhost"
                         <> help "Circus server host address" )
            <*> strOption ( long "circus-port" 
                         <> short 'P'
                         <> metavar "PORT" 
                         <> showDefault 
                         <> value "6007"
                         <> help "Circus server port" )
            <*> strOption ( long "ace" 
                         <> short 'i'
                         <> metavar "ID" 
                         <> showDefault 
                         <> value "op2"
                         <> help "ACE OP ID. One of op1, op2 (default), op8" )
            <*> strOption ( long "pdk" 
                         <> short 'p'
                         <> metavar "PDK" 
                         <> showDefault 
                         <> value "xh035"
                         <> help "ACE Backend. One of xh018, xh035 (default)" )
            <*> strOption ( long "space"
                         <> short 's'
                         <> metavar "SPACE"
                         <> showDefault
                         <> value "elec"
                         <> help "Design / Action space. elec (default) or geom" )
            <*> strOption ( long "var"
                         <> short 'v'
                         <> metavar "VARIANT"
                         <> showDefault
                         <> value "0"
                         <> help "Circus Environment Variant. 0 = goal (default), 1 = non-goal" )
            <*> strOption ( long "path"
                         <> short 'f'
                         <> metavar "FILE"
                         <> showDefault
                         <> value "./models"
                         <> help "Base Path for Model Checkpoint. Default is ./models" )
            <*> strOption ( long "tracking-host"
                         <> short 'T'
                         <> metavar "HOST"
                         <> showDefault
                         <> value "localhost"
                         <> help "MLFlow tracking server host address" )
            <*> strOption ( long "tracking-port"
                         <> short 'R'
                         <> metavar "PORT"
                         <> showDefault
                         <> value "6008"
                         <> help "MLFlow tracking server port" )
            <*> strOption ( long "mode"
                         <> short 'm'
                         <> metavar "MODE"
                         <> showDefault
                         <> value "Train"
                         <> help "Run Mode. One of Train (default), Continue, Evaluate" )
